### DISCUSSION TITLE
Wikilink Auto-completion Tweak

### DISCUSSION THREAD

**Discussion by yttriumz** [foambubble/foam#1153] [Ideas]
Could we add an option to auto-complete the wikilink with *full path* or *include its parent folder*?

For example, in the following note base:

```text
notes
└── OS
    ├── cross-platform
    │   ├── starship.md
    │   └── wezterm.md
    └── Linux
        └── openSUSE
            └── Tumbleweed
                └── installation.md
```

If we want to create a wikilink to `Tumbleweed/installation.md` in `starship.md`, the auto-completed wikilink will be `[[installation]]`.

However, later we may add notes for Ubuntu in our note base, and the note base becomes the following:

```text
notes
└── OS
    ├── cross-platform
    │   ├── starship.md
    │   └── wezterm.md
    └── Linux
        ├── openSUSE
        │   └── Tumbleweed
        │       └── installation.md
        └── Ubuntu 20.04
            └── installation.md
```

Then the wikilink `[[installation]]` in `starship.md` will cause trouble because it doesn't differentiate between `openSUSE/Tumbleweed/installation.md` and `Ubuntu 20.04/installation.md`.

So I guess we can change the default wikilink to one of the following:

- `[[notes/Linux/openSUSE/Tumbleweed/installation]]`
- `[[Tumbleweed/installation]]`

Yes Foam can detect ambiguous wikilinks for us, but could we add this option so that we don't have to frequently fix ambiguous links in our former notes? 

**Update**: The intention of this discussion is to reduce the chance of ambiguous links occurring. But it would be better if Foam's overall refactoring ability gets improved.

**Comment by riccardoferretti** [comment]
Have you looked at Foam Identifiers? https://github.com/foambubble/foam/blob/745acbabd3c4af7ba65975812445692565410946/docs/dev/proposals/wikilinks-in-foam.md

**Reply by yttriumz** [reply]
Thanks for replying. I checked the doc, but I'm not sure what you mean...

Yes I know that Foam can detect ambiguity, but Foam does not actively prompt the user for correction. If we have an ambiguous wikilink now (but it was not ambiguous in the past), we will never know until we open the source note containing it in VSCode. But I guess we might prefer to fix the ambiguity as soon as it occurs.

**Reply by riccardoferretti** [reply]
Ah ok, I think I understand now.

Basically, whenever a new file is created (and deleted?) we should check if it creates an ambiguity with any existing link, is this what you are suggesting?

**Reply by yttriumz** [reply]
Yes! But maybe not limited to file creation or deletion. Maybe whenever broken/ambiguous links are detected, there should be a prompt for the user's correction. For example, if I change a heading in a note, which is linked by another note, there will be a broken link because foam could not "refactor" that link.

BTW, although I'm new to foam, I really like it and have been using it a lot lately. But I also wish foam's refactoring ability could get improved to support folder and heading renaming.

**Reply by riccardoferretti** [reply]
Yeah the heading should be treated as a symbol to support refactoring (similar to how you'd rename a variable in JS). It's on the (long) todo list.
Currently diagnostics are only run on the open file, that was an arbitrary choice, I don't think it would be complicated to run it over the whole workspace.
But yeah I see what you mean, both for the sections and the folders, we could monitor more events/situations to inform the user of any inconsistency.
Can you please create (individual) items for these issues so they get tracked and tackled?

**Reply by yttriumz** [reply]
Issue created. Thanks for your work!
